### PR TITLE
Add collect() Aggregation Function

### PR DESCRIPTION
**Pull request by amansinghoriginal**
# Description

## Summary
This PR implements the `collect()` aggregation function in Drasi Core to solve the known issue where queries with multiple WITH clauses containing different grouping keys produce multiple partial results instead of properly aggregated results.

## Problem Solved
When using queries like:
```cypher
MATCH (p:products)-[:PRODUCT_TO_ORDER_ITEM]->(oi:orderItem),
      (r:reviews)-[:REVIEW_TO_PRODUCT]->(p)
WITH p, r, count(oi) as orderCount, avg(oi.quantity) as avgQuantity
WITH p, orderCount, avgQuantity, avg(r.rating) as avgRating, count(r) as reviewCount
RETURN p.productId, orderCount, avgQuantity, avgRating, reviewCount
```

The query would produce multiple partial results per product (e.g., 7 results for one product, 4 for another) instead of one properly aggregated result per product.

## Solution
The `collect()` function aggregates values into lists, avoiding the cartesian product issue:

```cypher
MATCH (p:products)-[:PRODUCT_TO_ORDER_ITEM]->(oi:orderItem)
WITH p, collect(oi.quantity) as order_quantities
OPTIONAL MATCH (r:reviews)-[:REVIEW_TO_PRODUCT]->(p)
WITH p, order_quantities, collect(r.rating) as review_ratings
RETURN p.productId, size(order_quantities) as orderCount, 
       size(review_ratings) as reviewCount
```

## Implementation Details
- Implements the `AggregatingFunction` trait with proper `apply()`, `revert()`, and `snapshot()` methods
- Handles null values correctly (ignores them)
- Supports collecting any variable value type
- Works seamlessly with all storage backends (in-memory, Garnet, RocksDB) without requiring backend-specific changes

## Testing
- 9 unit tests covering edge cases
- 7 shared integration tests
  - Solution validation with collect()
  - Filtering with WHERE clauses
  - Collecting complex objects
  - Multiple collect() in same WITH clause
  - Mixed data types
- Tests verified on all storage backends

### CODE REVIEW

File: core/src/evaluation/functions/aggregation/collect.rs [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+                return Err(FunctionError {
+                    function_name: "Collect".to_string(),
+                    error: FunctionEvaluationError::CorruptData,
+                })
+            }
+        };
+
+        // For revert, we need to remove the value from the list
+        // This is tricky because we need to find and remove the exact value
+        // For now, we'll remove the first occurrence
```
Comment (ruokun-niu) [inline core/src/evaluation/functions/aggregation/collect.rs:111]: What would happen if we have duplicate values? I am just curious about what is the expected behavior

Comment (amansinghoriginal) [reply core/src/evaluation/functions/aggregation/collect.rs:111]: If there are duplicates, it will remove the first occurrence (the one with lowest index). This effectively treats duplicates with FIFO semantics.

Comment (amansinghoriginal) [reply core/src/evaluation/functions/aggregation/collect.rs:111]: Both the OpenCypher and GQL standards state that COLLECT does not guarantee order. Aggregations into lists are generally considered to be from an unordered multiset.

In a streaming system like Drasi, "order" could means arrival order.

So if the collected list contains [A, B, A, C]) and a revert operation is triggered for A, the implementation will remove only the first occurrence of A it finds. The resulting list would be [B, A, C].

Comment (amansinghoriginal) [reply core/src/evaluation/functions/aggregation/collect.rs:111]: In Drasi's terms, if we're collecting different entities (nodes) in the graph - they will anyways be unique because of Element ID.

Comment (ruokun-niu) [reply core/src/evaluation/functions/aggregation/collect.rs:111]: Makes sense. Thanks for the detailed explanation

**Review by danielgerlag — APPROVED** [review]

**Review by ruokun-niu — APPROVED** [review]
